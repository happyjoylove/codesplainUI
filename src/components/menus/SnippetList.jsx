import React, { Component, PropTypes } from 'react';
import { MenuItem } from 'material-ui';
import ArrowDropRight from 'material-ui/svg-icons/navigation-arrow-drop-right';

import CustomPropTypes from '../../util/custom-prop-types';

const makeMenuItems = (snippetOwner, snippetsList, onClick) => {
  if (Object.keys(snippetsList).length === 0) {
    return null;
  }
  return Object.keys(snippetsList).map(snippetKey => (
    <MenuItem
      key={snippetKey}
      onTouchTap={() => onClick(snippetOwner, snippetKey)}
      primaryText={snippetsList[snippetKey].snippetTitle}
    />
  ));
};

class SnippetList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      focusState: 'none',
    };
    this.handleOnTouchTap = this.handleOnTouchTap.bind(this);
  }

  componentWillUnmount() {
    this.setState({ focusState: 'none' });
  }

  handleOnTouchTap() {
    this.setState({
      focusState: 'focused',
    });
  }

  render() {
    const {
      onClick,
      primaryText,
      snippetOwner,
      snippetsList,
    } = this.props;
    const {
      focusState,
    } = this.state;
    const menuItems = makeMenuItems(snippetOwner, snippetsList, onClick);
    return (
      <MenuItem
        disabled={!menuItems}
        focusState={focusState}
        menuItems={menuItems}
        onTouchTap={this.handleOnTouchTap}
        primaryText={primaryText}
        rightIcon={<ArrowDropRight />}
      />
    );
  }
}

SnippetList.propTypes = {
  onClick: PropTypes.func.isRequired,
  primaryText: PropTypes.string.isRequired,
  snippetsList: CustomPropTypes.snippets.isRequired,
  snippetOwner: PropTypes.string.isRequired,
};

export default SnippetList;
