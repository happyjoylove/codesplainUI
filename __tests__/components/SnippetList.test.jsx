import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import SnippetList from '../../src/components/menus/SnippetList';

const defaultProps = {
  onClick: jest.fn(),
  primaryText: '',
  snippetsList: {},
  snippetOwner: 'username',
};

describe('<SnippetList />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });

  describe('snapshot tests', () => {
    it('matches when no snippets are passed', () => {
      const wrapper = shallowWithContext(
        <SnippetList
          {...defaultProps}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('matches when snippets are passed', () => {
      const snippetsList = {
        foo: { snippetTitle: 'bar' },
        baz: { snippetTitle: 'qux' },
      };
      const wrapper = shallowWithContext(
        <SnippetList
          {...defaultProps}
          snippetOwner="username"
          snippetsList={snippetsList}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });
});
