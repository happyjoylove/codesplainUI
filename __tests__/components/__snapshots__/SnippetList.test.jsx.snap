exports[`<SnippetList /> snapshot tests matches when no snippets are passed 1`] = `
<MenuItem
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  checked={false}
  desktop={false}
  disabled={true}
  focusState="none"
  insetChildren={false}
  menuItems={null}
  onTouchTap={[Function]}
  primaryText=""
  rightIcon={<NavigationArrowDropRight />} />
`;

exports[`<SnippetList /> snapshot tests matches when snippets are passed 1`] = `
<MenuItem
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  checked={false}
  desktop={false}
  disabled={false}
  focusState="none"
  insetChildren={false}
  menuItems={
    Array [
      <MenuItem
        anchorOrigin={
            Object {
                "horizontal": "right",
                "vertical": "top",
              }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        onTouchTap={[Function]}
        primaryText="bar" />,
      <MenuItem
        anchorOrigin={
            Object {
                "horizontal": "right",
                "vertical": "top",
              }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        onTouchTap={[Function]}
        primaryText="qux" />,
    ]
  }
  onTouchTap={[Function]}
  primaryText=""
  rightIcon={<NavigationArrowDropRight />} />
`;
